File: java15/src/main/java/ast/ABranch.java
package ast;

import java.util.List;

import static java.util.stream.Collectors.toList;

public record ABranch(List<ATree> value) implements ATree {

    public ALeaf evalAST(ast.State state) {
        final var head = value.get(0);
        final var function = head.evalAST(state);
        final List<ALeaf> collect = value.subList(1, value.size()).stream().map((a) -> a.evalAST(state)).collect(toList());
        return function.applyAST(collect, state);
    }

    public String toString() {
        return "[L " + String.join(" ", value.stream().map(Object::toString).collect(toList())) + " R]";
    }

    @Override
    public ALeaf applyAST(List<ALeaf> args, ast.State state) {
        return null;
    }

//    @Override
//    public String toString() {
//        final StringBuilder xx = new StringBuilder();
//        xx.append("[");
//        for (ATree x : value) {
//            xx.append(x);
//            xx.append(" ");
//        }
//        xx.append("]");
//        return xx.toString();
//    }

}

File: kotlin/src/ast/ABranch.kt
package ast

import java.util.function.Function
import java.util.stream.Collectors

data class ABranch(val value: List<ATree>) : ATree {

    override fun evalAST(state: State?): ALeaf? {
        val head: ATree = value.get(0)
        val function = head.evalAST(state)
        val collect: List<ALeaf?> = value.subList(1, value.size).stream()
            .map(Function { a: ATree -> a.evalAST(state) })
            .collect(Collectors.toList())
        return function?.applyAST(collect, state)
    }

    override fun toString(): String {
        val collect = value.stream()
            .map<String>(Function { obj: ATree -> obj.toString() })
            .collect(Collectors.toList())
        return "[L " + java.lang.String.join(" ", collect) + " R]"
    }

    override fun applyAST(args: List<ALeaf?>?, st: State?): ALeaf {
        return AString("nothing")
    } //    @Override
    //    public String toString() {
    //        final StringBuilder xx = new StringBuilder();
    //        xx.append("[");
    //        for (ATree x : value) {
    //            xx.append(x);
    //            xx.append(" ");
    //        }
    //        xx.append("]");
    //        return xx.toString();
    //    }
}